app/models/immunisation_import.rb

```diff
@@ -21,16 +21,32 @@
 #  fk_rails_...  (user_id => users.id)
 #
 class ImmunisationImport < ApplicationRecord
+  attr_accessor :csv_is_malformed, :data
+
   belongs_to :user
 
+  EXPECTED_HEADERS = %w[DATE_OF_VACCINATION].freeze
+
   validates :csv, presence: true
+  validate :csv_is_valid
+  validate :csv_has_records
+  validate :headers_are_valid
 
   def csv=(value)
     super(value.respond_to?(:read) ? value.read : value)
   end
 
+  def load_data!
+    return if invalid?
+
+    self.data ||= CSV.parse(csv, headers: true, skip_blanks: true)
+  rescue CSV::MalformedCSVError
+    self.csv_is_malformed = true
+  ensure
+    csv.close if csv.respond_to?(:close)
+  end
+
   def process!(patient_session:)
```

Comment:
Should this `raise` or something similar if we haven't previously called `load_data!`, i.e. `data` is `nil`?

Reply:
What about calling `load_data!` if `data.nil?`?

Reply:
Yeah that would work too, although it would still then need to check the data was loaded correctly.